client/asset/btc/spv.go

```diff
@@ -0,0 +1,1712 @@
+	GetBlock(blockHash chainhash.Hash, options ...neutrino.QueryOption) (*btcutil.Block, error)
+	Stop() error
+}
+
+var _ neutrinoService = (*neutrino.ChainService)(nil)
+
+// createSPVWallet creates a new SPV wallet.
+func createSPVWallet(privPass []byte, seed []byte, dbDir string, log dex.Logger, net *chaincfg.Params) error {
+	netDir := filepath.Join(dbDir, net.Name)
+
+	if err := logNeutrino(netDir); err != nil {
+		log.Warnf("error initializing btcwallet+neutrino logging: %v", err)
+	}
+
+	logDir := filepath.Join(netDir, logDirName)
+	err := os.MkdirAll(logDir, 0744)
+	if err != nil {
+		return fmt.Errorf("error creating wallet directories: %v", err)
+	}
+
+	loader := wallet.NewLoader(net, netDir, true, 60*time.Second, 250)
+
+	pubPass := []byte(wallet.InsecurePubPassphrase)
+	_, err = loader.CreateNewWallet(pubPass, privPass, seed, time.Now())
+	if err != nil {
+		return fmt.Errorf("CreateNewWallet error: %w", err)
+	}
+
+	bailOnWallet := func() {
+		if err := loader.UnloadWallet(); err != nil {
+			log.Errorf("Error unloading wallet after createSPVWallet error: %v", err)
+		}
+	}
+
+	neutrinoDBPath := filepath.Join(netDir, neutrinoDBName)
+	db, err := walletdb.Create("bdb", neutrinoDBPath, true, 5*time.Second)
+	if err != nil {
+		bailOnWallet()
+		return fmt.Errorf("unable to create wallet db at %q: %v", neutrinoDBPath, err)
+	}
+	if err = db.Close(); err != nil {
+		bailOnWallet()
+		return fmt.Errorf("error closing newly created wallet database: %w", err)
+	}
+
+	if err := loader.UnloadWallet(); err != nil {
+		return fmt.Errorf("error unloading wallet: %w", err)
+	}
+
+	return nil
+}
+
+var (
+	// rotatorMtx protects the rotating file logger trackers.
+	rotatorMtx sync.Mutex
+	// loggerCount is a count of the file logger users. We'll Close the logger
+	// when the count goes to zero.
+	loggerCount uint32
+	// sharedRotator is the active rotating file logger. It must be Close'd when
+	// the last user is done logging.
+	sharedRotator *rotator.Rotator
+)
+
+// logRotator initializes a rotating file logger.
+func logRotator(netDir string) (*rotator.Rotator, error) {
+	const maxLogRolls = 8
+	logDir := filepath.Join(netDir, logDirName)
+	if err := os.MkdirAll(logDir, 0744); err != nil {
+		return nil, fmt.Errorf("error creating log directory: %w", err)
+	}
+
+	logFilename := filepath.Join(logDir, "neutrino.log")
+	return rotator.New(logFilename, 32*1024, false, maxLogRolls)
+}
+
+// logNeutrino initializes logging in the neutrino + wallet packages. Logging
+// only has to be initialized once, and the rotating file logger must be Close'd
+// at some point, so a package level counter is used to synchronize feeders.
+// The unlogNeutrino function should be invoked by callers of logNeutrino when
+// they no longer need logging.
+func logNeutrino(netDir string) error {
+	rotatorMtx.Lock()
+	defer rotatorMtx.Unlock()
+	if loggerCount > 0 {
+		return nil
```

Comment:
I think you still want to increment here.

```go
if loggerCount > 0 {
	loggerCount++
	return nil
}
```

seems like the only place not to increment is with a non-nil return setting up the logger.

Reply:
I confirmed that the remaining log races go away with this, but then realized that in `createSPVWallet` (the other place where `logNeutrino` is called), we need a `defer unlogNeutrino()` too otherwise it never decrements to zero to close the log rotator.

Sadly, this reveals a problem: https://github.com/lightninglabs/neutrino/blob/51686db787e01a3709b1583af3e20e916811d585/neutrino.go#L1136-L1137

The time drain on this logger is unfortunate, and it's looking like only ever calling UseLogger _once_ (somehow), even if we are synchronizing the best we can, is the only actual solution.  I'm not even sure a dumb Sleep would do it, but perhaps it's worth doing that just to save ourselves further wasted time.